libraries/AP_OSD/AP_OSD_ParamScreen.cpp

```diff
@@ -0,0 +1,449 @@
+extern const AP_HAL::HAL& hal;
+
+// constructor
+AP_OSD_ParamScreen::AP_OSD_ParamScreen()
+{
+}
+
+void AP_OSD_ParamScreen::draw_parameter(uint8_t number, uint8_t x, uint8_t y)
+{
+    AP_OSD_ParamSetting& setting = params[number-1];
+    setting.update();
+
+    ap_var_type type = setting.type;
+    AP_Param* p =  setting.param;
+    if (p != nullptr) {
+        const bool selected = number == _selected_param;
+        uint16_t value_pos = strlen(setting._name) + 3;
+        backend->write(x, y, selected && !_modify_param, "%s:", setting._name);
+
+        const bool blink = selected && _modify_param;
+        switch (type) {
+        case AP_PARAM_INT8:
+            backend->write(value_pos, y, blink, "%hhd", ((AP_Int8*)p)->get());
+            break;
+        case AP_PARAM_INT16:
+            backend->write(value_pos, y, blink, "%hd", ((AP_Int16*)p)->get());
+            break;
+        case AP_PARAM_INT32:
+            backend->write(value_pos, y, blink, "%d", ((AP_Int32*)p)->get());
+            break;
+        case AP_PARAM_FLOAT:
+            backend->write(value_pos, y, blink, "%.3f", ((AP_Float*)p)->get());
+            break;
+        case AP_PARAM_VECTOR3F:
+        {
+            Vector3f vec = ((AP_Vector3f*)p)->get();
+            backend->write(x, y, blink, "[%.3f,%.3f,%.3f]", vec.x, vec.y, vec.z);
+        }
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+// modify the selected parameter number
+void AP_OSD_ParamScreen::modify_parameter(uint8_t number, Event ev)
+{
+    const AP_OSD_ParamSetting& setting = params[number-1];
+    AP_Param* p = setting.param;
+
+    if (p->is_read_only()) {
+        return;
+    }
+
+    int32_t incr = (ev == Event::MENU_DOWN) ? -1 : 1;
+
+    if (p != nullptr) {
+        switch (setting.type) {
+        // there is no way to validate the ranges, so as a rough guess prevent
+        // integer types going below -1;
+        case AP_PARAM_INT8: 
+        {
+            AP_Int8* param = (AP_Int8*)p;
+            param->set(constrain_int16(param->get() + incr, -1, 127));
+        }
+            break;
+        case AP_PARAM_INT16:
+        {
+            AP_Int16* param = (AP_Int16*)p;
+            param->set(constrain_int16(param->get() + incr, -1, 32767));
+        }
+            break;
+        case AP_PARAM_INT32:
+        {
+            AP_Int32* param = (AP_Int32*)p;
+            param->set(constrain_int32(param->get() + incr, -1, INT_MAX));
+        }
+            break;
+        case AP_PARAM_FLOAT:
+        {
+            AP_Float* param = (AP_Float*)p;
+            param->set(param->get() + incr * 0.001f);
+        }
+            break;
+        case AP_PARAM_VECTOR3F:
+        {
+            AP_Vector3f* param = (AP_Vector3f*)p;
+            Vector3f vec = param->get();
+            vec.x += incr * 0.001f;
+            vec.y += incr * 0.001f;
+            vec.z += incr * 0.001f;
+            param->set(vec);
+        }
+            break;
+        default:
```

Comment:
Omit the default and just break-and-do-nothing for types you don't understand.  There's not that many :-)

Reply:
Fixed